rfcs/0004-range-queries.md

```diff
@@ -0,0 +1,217 @@
+may `seek` using the returned iterator.
+
+```rust
+impl Db {
+    /// Scan a range of keys.
+    /// 
+    /// returns a `DbIterator`
+    pub async fn scan(
+        &self,
+        range: std::ops::Range<&[u8]>
+    ) -> Result<DbIterator, SlateDbError>;
+}
+
+impl DbIterator {
+    /// Get the next record in the scan.
+    /// 
+    /// returns Ok(None) when the scan is complete
+    /// returns Err(InvalidatedIterator) if the iterator has been invalidated
+    ///  due to an underlying error
+    pub async fn next(
+        &mut self,
+    ) -> Result<Option(DbRecord), SlateDbError>;
+
+    /// Seek to a key ahead of the last key returned from the iterator or
+    /// the lower range bound if no records have yet been returned.
+    /// 
+    /// returns Ok(()) if the position is successfully advanced
+    /// returns SlateDbError::InvalidArgument if `lower_bound` is `Unbounded`
+    /// returns SlateDbError::InvalidArgument if the key is comes before the
+    ///  current iterator position.
+    /// returns Err(InvalidatedIterator) if the iterator has been invalidated
+    ///  due to an underlying error
+    pub async fn seek(
+        &mut self,
+        lower_bound: Bound<&[u8]>
+    ) -> Result<(), SlateDbError>;
+}
+
+struct DbRecord {
+    key: Bytes,
+    value: Bytes,
+}
+```
+
+Note that the `DbIterator` may be invalidated internally if any of its referenced
+resources must be reclaimed. In this case, a call to `DbIterator::next` or 
+`DbIterator::scan` will return `SlateDb::InvalidatedError`. This is discussed in 
+more detail below.
+
+We will also provide `Db::scan_with_options`, which allows the user to specify the
+query isolation through a `ScanOptions` struct. Similar to `get_with_options`, the
+two isolation levels that will be supported are `Committed` and `Uncommitted`.
+The default when `scan` is called without explicit options will be `Committed`.
+
+```rust
+impl Db {
+    /// Scan a range of keys with the provided options.
+    ///
+    /// returns a `DbIterator`
+    pub async fn scan_with_options(
+        &self,
+        range: &DbRange,
+        options: &ScanOptions,
+    ) -> Result<DbIterator, SlateDbError>;
+}
+
+pub struct ScanOptions {
+    /// The read commit level for read operations
+    pub read_level: ReadLevel,
+    /// The number of bytes to read ahead
+    pub read_ahead_size: usize,
+    /// Whether or not fetched blocks should be cached
+    pub cache_blocks: bool
+}
+```
+
+## Implementation
+
+A scan must iterate every table which may have keys within the query range.
+This includes tables in memory which which have not been flushed and L0/SR tables 
+which may need to be loaded from object storage. Internally, we will rely on 
+iterators which sort and merge these tables. The latest records will have precedence following
+the same rules that a normal `get` query follows.
+
+The existing `MergeIterator` used for compaction provides much of the necessary
```

Comment:
Can you expand this section a bit more? Would `Dbiterator` implementation create a `MergeIterator` with all the relevant files first, and then use MergeIterator?

Reply:
I will add some detail. I expected to filter the set of tables when building the iterator and free them as they are used. Let me know if there is any reason to do otherwise.